src/e2ee/matrixKeyProvider.ts

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2023 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { BaseKeyProvider, createKeyMaterialFromString } from "livekit-client";
+import {
+  MatrixRTCSession,
+  MatrixRTCSessionEvent,
+} from "matrix-js-sdk/src/matrixrtc/MatrixRTCSession";
+
+export class MatrixKeyProvider extends BaseKeyProvider {
+  private rtcSession?: MatrixRTCSession;
+
+  public setRTCSession(rtcSession: MatrixRTCSession) {
+    if (this.rtcSession) {
+      this.rtcSession.off(
+        MatrixRTCSessionEvent.EncryptionKeyChanged,
+        this.onEncryptionKeyChanged
+      );
+    }
+
+    this.rtcSession = rtcSession;
+
+    this.rtcSession.on(
+      MatrixRTCSessionEvent.EncryptionKeyChanged,
+      this.onEncryptionKeyChanged
+    );
+
+    for (const [
+      participant,
+      encryptionKey,
+    ] of this.rtcSession.encryptionKeys.entries()) {
+      this.onEncryptionKeyChanged(
```

Comment:
Some comments here wouldn't go amiss: I'm not immediately seeing why the RTCSession changing would obviously imply a key change.

Reply:
Not necessarily but the new session could be aware of keys of which the old session wasn't. Also, if we end up linking the encryption key events to individual calls as you suggested in the other PR (which makes sense), then I assume the keys would change with the session (that is if I understand correctly what the class represents - i.e. there would be multiple instances of this class per room if there were two call ids being used in that room)